wysteria/middleware/impl_nats.py

```diff
@@ -85,12 +59,68 @@ def retry_func(*args, **kwargs):
         for count in range(0, NATS_MSG_RETRIES + 1):
             try:
                 return func(*args, **kwargs)
-            except Exception as e:
+            except (RequestTimeoutError, Empty) as e:
                 if count >= NATS_MSG_RETRIES:
                     raise
     return retry_func
 
 
+class _TornadoNats(threading.Thread):
+    """
+    Tiny class to handle queuing requests through tornado.
+    """
+    _MAX_RECONNECTS = 10
+
+    def __init__(self, url):
+        threading.Thread.__init__(self)
+        self._conn = None
+        self._url = url
+        self._outgoing = Queue()
+        self._running = False
+
+    @tornado.gen.coroutine
+    def main(self):
+        """Connect to remote host(s)
+
+        Raises:
+            Exception if unable to establish connection to remote host(s)
+        """
+        self._conn = Client()
+        yield self._conn.connect(
+            servers=[self._url],
+            allow_reconnect=True,
+            max_reconnect_attempts=self._MAX_RECONNECTS
+        )
+        while self._running:
+            while not self._outgoing.empty():
+                reply_queue, key, data = self._outgoing.get()
+                result = yield self._conn.timed_request(key, data)
+                reply_queue.put(result.data)
+
```

Comment:
If the requestor used a timeout that is shorter than the time it takes for this request to complete, they will stop pulling from their reply queue. This would leave the client loop deadlocked as no one will ever drain the reply.

Reply:
True .. I could put a min limit on this?

Reply:
Hmm, maybe it won't block, because you are using a buffered `Queue`. As long as the buffer size is always matched to the number of times it can retry, there technically should always be a slot that wont block. You could maybe switch it to `put_nowait()` so that it really would never block for any reason, regardless of the buffering size. Just a thought.

Reply:
Put no wait seems like a good plan .. it's probably better that we avoid any potential deadlock chances ?